src/PowerShellEditorServices/Services/PowerShell/Execution/PipelineThreadExecutor.cs

```diff
@@ -0,0 +1,284 @@
+
+        private readonly ILogger _logger;
+
+        private readonly EditorServicesConsolePSHost _psesHost;
+
+        private readonly IReadLineProvider _readLineProvider;
+
+        private readonly HostStartupInfo _hostInfo;
+
+        private readonly BlockingCollection<ISynchronousTask> _executionQueue;
+
+        private readonly CancellationTokenSource _consumerThreadCancellationSource;
+
+        private readonly Thread _pipelineThread; 
+
+        private readonly CancellationContext _loopCancellationContext;
+
+        private readonly CancellationContext _commandCancellationContext;
+
+        private readonly ReaderWriterLockSlim _taskProcessingLock;
+
+        private bool _runIdleLoop;
+
+        public PipelineThreadExecutor(
+            ILoggerFactory loggerFactory,
+            HostStartupInfo hostInfo,
+            EditorServicesConsolePSHost psesHost,
+            IReadLineProvider readLineProvider)
+        {
+            _logger = loggerFactory.CreateLogger<PipelineThreadExecutor>();
+            _hostInfo = hostInfo;
+            _psesHost = psesHost;
+            _readLineProvider = readLineProvider;
+            _consumerThreadCancellationSource = new CancellationTokenSource();
+            _executionQueue = new BlockingCollection<ISynchronousTask>();
+            _loopCancellationContext = new CancellationContext();
+            _commandCancellationContext = new CancellationContext();
+            _taskProcessingLock = new ReaderWriterLockSlim(LockRecursionPolicy.SupportsRecursion);
+
+            _pipelineThread = new Thread(Run)
+            {
+                Name = "PSES Execution Service Thread",
+            };
+            _pipelineThread.SetApartmentState(ApartmentState.STA);
+        }
+
+        public bool IsExiting { get; set; }
+
+        public Task<TResult> QueueTask<TResult>(SynchronousTask<TResult> synchronousTask)
+        {
+            _executionQueue.Add(synchronousTask);
+            return synchronousTask.Task;
+        }
+
+        public void Start()
+        {
+            _pipelineThread.Start();
+        }
+
+        public void Stop()
+        {
+            _consumerThreadCancellationSource.Cancel();
+            _pipelineThread.Join();
+        }
+
+        public void CancelCurrentTask()
+        {
+            _commandCancellationContext.CancelCurrentTask();
+        }
+
+        public void Dispose()
+        {
+            Stop();
+        }
+
+        public IDisposable TakeTaskWriterLock()
+        {
+            return TaskProcessingWriterLockLifetime.TakeLock(_taskProcessingLock);
+        }
+
+        private void Run()
+        {
+            _psesHost.PushInitialPowerShell();
+            // We need to override the idle handler here,
+            // since readline will be overridden when the initial Powershell runspace is instantiated above
+            _readLineProvider.ReadLine.TryOverrideIdleHandler(OnPowerShellIdle);
```

Comment:
ReadLine is null here in my testing, causing a NullReferenceException. _readLineProvider is not null. I changed this to `_readLineProvider.ReadLine?.TryOverrideIdleHandler(OnPowerShellIdle);` and it worked as expected.